### PR TITLE
New ASTCanopy Build System

### LINKED ISSUE #52: Current Build System Does not Fit into Conda Build System
In #45, we did a lot of manual wrangling for `libastcanopy.so`. Fundamentally this is because scikit-build-core doesn't know how to handle shared libraries when the project (pylibastcanopy) also implicitly builds another C++ project (libastcanopy), and depends on it. Locally this is all fine because previously this was installed to `$ENV(CONDA_PREFIX)/lib` in cmakelists. This is non-standard behavior. Cmake is meant to perform relocatable install and no specific paths to `lib/` should be specified. Preferably libastcanopy is installed in a separate cmake process, and `pylibastcanopy` "depends on" `libastcanopy` using `find_package` in its own cmake. This way conda build will recognize both shared libraries and deliver them in proper environment.

### PR DESCRIPTION
**Pull request by isVoid**
This PR changes astcanopy build system to the following:
- Use cmake / other build tools to build libastcanopy
- Change pylibastcanopy's library discovery via `find_package`
- Undo some custom build wrangling in #45 

Additionally, this PR restructures the cpp project layout following  C++ header layout convention (`<ast_canopy/ast_canopy.hpp>`).

This PR also removes `cuda-tookit` dependency for `ast_canopy`. Partially contributing to #53.

One question left unanswered is that ast-canopy recipe explicitly needs `{{ compiler('c') }}` in run/host. While `cuda-nvcc` already depends on some host compilers, nvcc needs the `gcc` binary to correctly function, and `gcc_linux-64` package doesn't provide this binary. This makes me feel that `cuda-nvcc` package recipe is incomplete. But I'm not sure, and it should not be blocking this PR. See [this CI job](https://github.com/NVIDIA/numbast/actions/runs/9502540686/job/26190770495?pr=54) for more detail.

Fixes #52

### CODE REVIEW

File: conda/recipes/ast_canopy/build.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+BUILD_DIR=$SRC_DIR/cpp/build
+CPP_DIR=$SRC_DIR/cpp
+
+echo SRC_DIR=$SRC_DIR
+echo CPP_DIR=$CPP_DIR
+echo BUILD_DIR=$BUILD_DIR
+
+export CMAKE_INSTALL_PREFIX=$PREFIX
+export CMAKE_CXX_STANDARD=11
+
```
Comment (jakirkham) [inline conda/recipes/ast_canopy/build.sh:23]: While some of these work as environment variables, would recommend just passing these as config values to the `cmake` call (as shown)

```suggestion
```

**Review by jakirkham — COMMENTED** [review]
Thanks Michael! 🙏 

Had a few suggestions below

Also think we may need `llvmdev` & `llvm` in `requirements/host`. Not sure what version we want, but would recommend pinning a version (maybe something that matches Numba to start? So 14 atm)

**Comment by isVoid** [comment]
[a6e99d1](https://github.com/NVIDIA/numbast/pull/54/commits/a6e99d17aa410dc9e67d250c9a3a11dda4f831a8) passes all CI tests. I'm now experimenting removing host compiler explicit dependency from conda meta.

File: conda/recipes/ast_canopy/meta.yaml [hunk 2]
```diff
@@ -40,15 +33,16 @@ requirements:
     - python
     - pip
     - scikit-build-core
+    - llvmdev >=14
+    - llvm >=14
   run:
     - {{ compiler('c') }}
     - {{ compiler('cxx') }}
```
Comment (jakirkham) [inline conda/recipes/ast_canopy/meta.yaml:40]:
```suggestion
    - {{ compiler('cuda') }}
```

File: conda/recipes/ast_canopy/meta.yaml [hunk 3]
```diff
@@ -40,15 +33,16 @@ requirements:
...
     - python
     - numba >=0.58
-    - pynvjitlink >=0.2
     - {{ pin_compatible('cuda-version', min_pin='x', max_pin='x.x') }}
-    - cuda-toolkit {{ cuda_compiler_version }}
-    - cuda-python {{ cuda_compiler_version }}
+    - cuda-nvcc {{ cuda_compiler_version }}
+    - cuda-nvcc-impl {{ cuda_compiler_version }}
```
Comment (jakirkham) [inline conda/recipes/ast_canopy/meta.yaml:45]:
```suggestion
```

**Review by jakirkham — COMMENTED** [review]
Thanks Michael! 🙏

Do we need the CUDA compiler at runtime? If so, would suggest something like this instead

Happy to discuss further as needed

**Pull request merged**